# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [5]:
import pandas as pd
# Task 1a code here...
temperatures = pd.Series([72, 75, 68, 71, 73],
                         index = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"])
print(temperatures)

Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
dtype: int64


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [11]:
# Task 1b code here...
products = pd.DataFrame({"product": ["Widget", "Gadget", "Doohickey"],
                         "price": [19.99, 24.99, 12.49],
                         "quantity": [100, 75, 150]})
print(products)
print(products.shape)

     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150
(3, 3)


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [12]:
# Task 1c code here...
print(products["price"])
print(products[["product", "quantity"]])

0    19.99
1    24.99
2    12.49
Name: price, dtype: float64
     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

1. The dictionary keys under "product", "price", and "quantity" become the column names and the values for each key became the data within each column, creating rows.
2. The df['price'] returns a series because it only pertains to one column, whereas df[['price']] will pass a group of column names. Pandas reads ['...'] as a 1 column string, a series, and reads [['...']] as multiple columns, a dataframe.
3. Using names rather than numbers when creating custom indexes would be more useful when organizing data that pertains to locations, data based on weekdays/time, questionaaire forms, etc. 

### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [6]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('./data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)
df_raw = pd.read_csv(csv_path)
df_raw.shape
df_raw.columns
df_raw.dtypes

# Check memory usage by column
print("Memory usage by column:")
print(df_raw.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

Memory usage by column:
Index                     132
ride_id               7300000
rideable_type         6950493
started_at            8000000
ended_at              8000000
start_station_name    7549653
start_station_id      5978313
end_station_name      7546619
end_station_id        5974035
start_lat              800000
start_lng              800000
end_lat                800000
end_lng                800000
member_casual         6300000
dtype: int64

Total memory usage:
63.70 MB


#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [23]:
# task 2b code here...
rides = pd.read_csv(csv_path,
                 dtype={'ride_id': 'category', 'rideable_type': 'category', 'start_station_name': 'category', 'member_casual': 'category', 'start_station_id': 'object', 'end_station_name': 'category'},
                 parse_dates=['started_at', 'ended_at'])
rides.info()
print("\nTotal memory usage:")
print(f"{rides.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  category      
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   category      
 5   start_station_id    89623 non-null   object        
 6   end_station_name    89485 non-null   category      
 7   end_station_id      89485 non-null   object        
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  category      
dtypes: category(5), datetime64[ns]

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [28]:
# task 2c code here...
date_minimum = rides['started_at'].min()
date_maximum = rides['started_at'].max()
print(f"min: {date_minimum}, max: {date_maximum}")
#count missing values
count_missing = rides.isna().sum()
print(f"missing values: {count_missing}")
#convert series
missing_value = count_missing.to_frame(name = 'count')
missing_value['percentage'] = (missing_value['count'] / len(rides))*100
print(f"Missing values: \n{missing_value}")

min: 2024-09-30 23:12:01.622000, max: 2024-10-31 23:54:02.851000
missing values: ride_id                   0
rideable_type             0
started_at                0
ended_at                  0
start_station_name    10377
start_station_id      10377
end_station_name      10515
end_station_id        10515
start_lat                 0
start_lng                 0
end_lat                  87
end_lng                  87
member_casual             0
dtype: int64
Missing values: 
                    count  percentage
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_station_name  10377      10.377
start_station_id    10377      10.377
end_station_name    10515      10.515
end_station_id      10515      10.515
start_lat               0       0.000
start_lng               0       0.000
end_lat                87       0.087
end_lng                87       0.087
member_casual           0      

#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [32]:
# task 2d code here...
#trip duration
rides['trip_duration_min'] = (rides['ended_at'] - rides['started_at']).dt.total_seconds() / 60
#display basic statistics
print('Trip duration basic stats:')
print(rides['trip_duration_min'].describe())
#show first few rows
print('First few rows:')
print(rides[['started_at', 'ended_at', 'trip_duration_min']].head())
#set started_at as DF index
rides = rides.set_index('started_at')
#verify
print(rides.index)
print(rides.head())

Trip duration basic stats:
count    100000.000000
mean         16.144576
std          52.922539
min           0.006533
25%           5.489271
50%           9.423592
75%          16.407171
max        1499.949717
Name: trip_duration_min, dtype: float64
First few rows:
               started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583
3 2024-09-30 23:42:11.207 2024-10-01 00:10:16.831          28.093733
4 2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          17.034933
DatetimeIndex(['2024-09-30 23:12:01.622000', '2024-09-30 23:19:25.409000',
               '2024-09-30 23:32:24.672000', '2024-09-30 23:42:11.207000',
               '2024-09-30 23:49:25.380000', '2024-09-30 23:49:40.016000',
               '2024-10-01 00:00:53.414000', '2024-10-01 00:05:44.954000',
               '202

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

1. Pandas assigned the type object, similar to python's string, to started_at and member_casual. These could be problematic for two reasons. First, started_at as an object means pandas cannot perform number analysis on the column. Second, member_casual as an object would increase memory and decrease performance.
2. While the station IDs seem random, they are most likely deterministic, which is similar to how git creates unique values.
3. Method chaining allows for multiple functions to be called at once, in the order it is written in. df.isna().sum() creates a dataframe in which missing/null values are highlighted and counted.
4. There could be a number of reasons for this. For instance, someone could have a multi-hop trip in which they travel within ports.To handle this, the rows with missing values should be flagged but not deleted as they are still real rides.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

Switching the types from object to category reduced the memeory usage strongly. The numbers are different from what is reported at the bottom of df.info() since df.info() shows a quick estimate of the memory usage and potentially may not account for multiple items. if data size is a concern, you should use df.memory_usage(deep=True).sum() since it is more accurate.
16GB x 30% = 4.8GB
16GB - 4.8 GB = 11.2 GB
Optimistic:
11.2GB / 5  = 2.24 GB
Conservative:
11.2GB / 10 = 1.12 GB
I would use the consevrative multiplier for more reliable work, especially because it keeps datasets around 1 GB.

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [36]:
# Task 3a code here...
members_w_electric = rides[(rides['member_casual'] == 'member') & 
                        (rides['rideable_type'] == 'electric_bike')]
num_mem_w_electric = len(members_w_electric)
print(f"number electric: {num_mem_w_electric}")
#percentage
percent = (num_mem_w_electric / len(rides))*100
print(f"percent: {percent}")
#num trips at 3 stations
stations = ["Streeter Dr & Grand Ave", 
                        "DuSable Lake Shore Dr & Monroe St", 
                        "Kingsbury St & Kinzie St"]
station_trips = rides[rides['start_station_name'].isin(stations)]
trip_count = len(station_trips)
print(f"Station trips: {trip_count}")

number electric: 33121
percent: 33.121
Station trips: 2702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [39]:
# Task 3b code here...
rides['is_weekend'] = rides.index.dayofweek.isin([5, 6])
#time
rides['time_of_day'] = "Midday"
hour = rides.index.hour
morning_mask = hour.isin([7, 8, 9])
rides.loc[morning_mask, 'time_of_day'] = "Morning Rush"
evening_mask = hour.isin([16, 17, 18])
rides.loc[evening_mask, 'time_of_day'] = "Evening Rush"
#print dat
print("Weekend trips distribution:")
print(rides['is_weekend'].value_counts())
print("Time of day distribution:")
print(rides['time_of_day'].value_counts())

Weekend trips distribution:
is_weekend
False    72785
True     27215
Name: count, dtype: int64
Time of day distribution:
time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64


#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [43]:
# Task 3c code here...
filtered = rides.query("member_casual == 'casual' and "
    "is_weekend == True and "
    "trip_duration_min > 20 and "
    "rideable_type == 'electric_bike'")
#count it
print(f"number trips: {len(filtered)}")
#percent
percentage = (len(filtered) / len(rides))*100
print(f"percentage: {percentage}")
#avg
avg = filtered['trip_duration_min'].mean()
print(f"average: {avg}")

number trips: 1501
percentage: 1.5010000000000001
average: 40.37352101932046


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [47]:
# Task 3d code here...
first_10 = rides.iloc[:10, rides.columns.get_indexer(['member_casual', 'rideable_type', 'trip_duration_min'])]
print(f"first 10: {first_10}")
#oct 15-17 trips
oct_trips = rides.loc['2024-10-15':'2024-10-17', ['member_casual', 'rideable_type', 'trip_duration_min']]
print(f"oct trips: {oct_trips.head(10)}")
#number trips
print(f"Num trips: {len(oct_trips)}")


first 10:                         member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-09-30 23:12:01.622        casual  electric_bike          67.984200
2024-09-30 23:19:25.409        casual  electric_bike          82.742067
2024-09-30 23:32:24.672        member   classic_bike          50.899583
2024-09-30 23:42:11.207        casual   classic_bike          28.093733
2024-09-30 23:49:25.380        member  electric_bike          17.034933
2024-09-30 23:49:40.016        member  electric_bike          13.009367
2024-10-01 00:00:53.414        member   classic_bike           2.598817
2024-10-01 00:05:44.954        member  electric_bike           0.013433
2024-10-01 00:06:12.035        member  electric_bike          10.472933
2024-10-01 00:10:03.646        member   classic_bike           7.825683
oct trips:                         member_casual  rideable_type  trip_duration_min
started_at                                 

#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

1. isin() is faster since it does not use multiple equal comparisons. It is also less error prone since multiple long chains will not need to be created.
2. Midday acts as a "null" category or a category in which all unassigned values are assigned to. If assigned in a different order or a default wasn't created, unassigned hours would have no value, or a NaN value.
3. query() should be used rather than boolean indexing when there are multiple conditions and where readability matters more. Otherwise, boolean indexing should be used when operations are too complex or perforamce is an absolute necessity.


#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

Using iloc[] and loc[] is better becuase they define the usage of their placement. FOr example, iloc[] means position based selection, and loc[] means label-based selection. This makes the code clearer and limits the error prone trait of using boolean indexing.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [50]:
# Task 4a code here...
hour_counts = rides.index.hour.value_counts()
peak_hour = hour_counts.idxmax()
peak_hour_trips = hour_counts.max()
print(f'peak hour: {peak_hour}')
print(f'peak hour trip number: {peak_hour_trips}')
#busiest day
day_count = rides.index.to_series().dt.day_name().value_counts()
busiest_day = day_count.idxmax()
busiest_day_trips = day_count.max()
print(f'busiest day: {busiest_day}')
print(f'busiest day trip number: {busiest_day_trips}')
#sorting
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
day_counts_sorted = day_count.reindex(day_order)
print(day_counts_sorted)


peak hour: 17
peak hour trip number: 10574
busiest day: Wednesday
busiest day trip number: 16513
started_at
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64


#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [56]:
# Task 4b code here...
avg_trip_duration_type = rides.groupby('member_casual')['trip_duration_min'].mean()
print(f'average trip duration: {avg_trip_duration_type}')
print(f"Casual group takes longer trips")
#bike duration
avg_duration_bike = rides.groupby('rideable_type')['trip_duration_min'].mean()
print(f'average bike duration: {avg_duration_bike}')
print('Classic bike takes longer')
#count w groupby
num_trips =  rides.groupby('member_casual').size()
print(f'number of trips: {num_trips}')
#count w value counts
num_trips_value_count = rides['member_casual'].value_counts()
print(f'number of trips with value count: {num_trips_value_count}')
print('While they do produce the same number per category, the output row order is switched')


average trip duration: member_casual
casual    23.978046
member    11.984493
Name: trip_duration_min, dtype: float64
Casual group takes longer trips
average bike duration: rideable_type
classic_bike     20.337410
electric_bike    12.033618
Name: trip_duration_min, dtype: float64
Classic bike takes longer
number of trips: member_casual
casual    34686
member    65314
dtype: int64
number of trips with value count: member_casual
member    65314
casual    34686
Name: count, dtype: int64
While they do produce the same number per category, the output row order is switched


C:\Users\halyn\AppData\Local\Temp\ipykernel_30200\1094518795.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_trip_duration_type = rides.groupby('member_casual')['trip_duration_min'].mean()
C:\Users\halyn\AppData\Local\Temp\ipykernel_30200\1094518795.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_duration_bike = rides.groupby('rideable_type')['trip_duration_min'].mean()
C:\Users\halyn\AppData\Local\Temp\ipykernel_30200\1094518795.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current beh

#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [62]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Task 4c code here...
# use the variable `output_file` as the filename for step 4
#1
weekend_electric = rides[(rides['is_weekend'] == True) & 
                         (rides['rideable_type'] == 'electric_bike')]
#2
weekend_elec_1k =  weekend_electric.iloc[:1000]
#3
reset =  weekend_elec_1k.reset_index()
#4 & #5
exported_col = ['started_at', 'ended_at', 'member_casual',
                'trip_duration_min', 'time_of_day']
reset[exported_col].to_csv(output_file, index = False)
#total weekend electric bike before 1000
print(f'total: {len(weekend_electric)}')

total: 13026


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

1. Grouby applies the "split-apply-combine" pattern. Split will separate groups based on the input in the groupby(parentheses). Apply will implement an operation like .mean() to the groups. And combine takes the results from the groups and merges them back together.
2. You would use value_counts() when you need to count values of one column and when you need a quick solution for counting. You would use groupby().size() when you need to find multiple statistics or when working with a more complex situation. You could not use value_counts() when more than one column needs to be counted.
3. We used reset_index() to convert the data back into a normal column so that it can be exported in the csv, otherwise it would be left out. If you exported using date-time index and used index = False, the started_at data would be lost as it would be considered "False" and not recognized.

#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [63]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides
    .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18]))]
    .loc[(rides['start_station_name'].str.contains('Lake Shore|Lakefront', 
                                                    case=False, 
                                                    na=False))]
    .loc[rides['trip_duration_min'] > 15]
    .sort_values('trip_duration_min', ascending=False)
    .head(1000)
    .reset_index()
    [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
      'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

Found 310 long rush-hour trips from lakefront stations

CSV file size: 40.57 KB
Pickle file size: 55.22 KB
Size difference: 14.64 KB

Load time comparison:
CSV:
1.22 ms ± 33.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Pickle:
508 μs ± 23.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name          category
end_station_name            category
member_casual               category
rideable_type               category
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


1. Since each operation is on its individual line, it makes the method chain clear to understand and read. it makes it easiet to debug as lines or methods can be commented out quickly to problem solve. With this you can also easily rearrange the chain if needed with this layout. This relates to seeing changes in code with git diff as it is also a line by line structure that allows for cleaner version history. Also, heavy chaining has multiple downsides like limited flexibility with adding code, performance issues, and making it harder to debug code as it is more complex.
2. While pickle maintains the dtypes, CSV converts the category and datetime dtypes to objects. This is significant as this will decrease memory space, decrease performance efficiency, and limit analysis via datetime types.
3. I would use pickle when I am using numerical or complex datatypes like datetime dytypes or categories because it preserves the dtype. ON the other hand, I would use CSV when I am not manipulating data that requires these dtypes. I would also use CSV if I was concerned about file size and sharing data.

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

1. a. I would say the biggest shift was realizing how easy it is to index a whole column using pandas. Pandas makes this much more simple as it can be done in even one line.

   b. Using loc[] and iloc[] was the most difficult for me, particularly becuase the names do not describe the function they are doing. So, I would easily confuse the two. To work through it, I realized the i in iloc[] meant integer, thus it uses integer positions. Knowing that loc[] did not have the 'i' meant it uses label-based positions.

2. a. Working with this CSV provided "real world" experience. Whereas the numpy arrays were already configured to start working with, this uncleanded csv file provided the hands-on experience to clean the data to enable proper analysis.

   b. Pandas allows for labeled data structures meaning, labeled rows and columns. It can also tolerate a mixture of data types whereas numpy can only tolerate uniform dtypes. As data in the real world is often a mixture of dtypes and pertains meaningful index and column names, Pandas is far more suited for data analysis.

3. a. I believe indexing using iloc[] and loc[] will be the most useful for me as it allows easily and quick integer and label-based indexing which is much easier to use in comparison to numpy's indexing. This will make comparing data columns easier in the future with my own data work.

   b. On a scale of 1-10 I feel that I comfortable to use pandas on my own work at a rating of 6. I feel like I definitely need more practice using pandas to lock in the spare details and functions within the library, however I do feel like I have the general basics of pandas down. Overall, to increase my confidence level in using pandas I woulld need to practice manipulating and analyzing data sets in my own time.

4. Feedback:
   Time spent: 8 hours
   - Most helpful part of the assignment: I feel that this homework gave students the opportunity to practice utilizing pandas in a "real world" style of analysis. It also gave many examples on the different ways to utilize pandas while ensuring the students understand the process and logic behind the code.
   - One specific improvement suggestion: It was a little too long for me. I feel that the content could have been spread into 3 problems rather than 4. Other than that, I feel that this homework was well-rounded and provided good guidance.